### PR TITLE
Adding HEVC/H265 FourCC support to MSMF video writer

### PR DESCRIPTION
**Pull request by lilohuang**
Adding HEVC/H265 FourCC support to MSMF video writer. I have verified it with my own video input stream, and it works well on my workstation.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by alalek** [comment]
> into opencv:master

`master` is not a valid branch for contribution: https://github.com/opencv/opencv/wiki/Branches
Please use 4.x (it is default on GitHub) - use "Edit" button near PR title to change the target branch

---

Please update videoio tests to check added new functionality for MSMF.

File: modules/videoio/src/cap_msmf.cpp [hunk 1]
```diff
@@ -2469,6 +2469,10 @@ const GUID CvVideoWriter_MSMF::FourCC2GUID(int fourcc)
 #endif
         case CV_FOURCC_MACRO('H', '2', '6', '4'):
                 return MFVideoFormat_H264; break;
+        case CV_FOURCC_MACRO('H', '2', '6', '5'):
+                return MFVideoFormat_H265;  break;
+        case CV_FOURCC_MACRO('H', 'E', 'V', 'C'):
+                return MFVideoFormat_HEVC;  break;
```
Comment (alalek) [inline modules/videoio/src/cap_msmf.cpp:2475]: There is a compilation problem with some Windows SDK versions.

> `error C2065: 'MFVideoFormat_H265': undeclared identifier`

Need a proper `#if` check/guard for using of these values.

Comment (lilohuang) [reply modules/videoio/src/cap_msmf.cpp:2475]: Will do. Thanks.

**Comment by alalek** [comment]
There is a problem on CI with H265/HEVC tests:

> cap_msmf.cpp:930 CvCapture_MSMF::initStream Failed to set mediaType (stream 0, (640x480 @ 25) MFVideoFormat_RGB32(codec not found)

Added this package:

> Microsoft.HEVCVideoExtension_2.0.60091.0_x64__8wekyb3d8bbwe.appx

**Comment by lilohuang** [comment]
> There is a problem on CI with H265/HEVC tests:
> 
> > cap_msmf.cpp:930 CvCapture_MSMF::initStream Failed to set mediaType (stream 0, (640x480 @ 25) MFVideoFormat_RGB32(codec not found)
> 
> Added this package:
> 
> > Microsoft.HEVCVideoExtension_2.0.60091.0_x64__8wekyb3d8bbwe.appx
> 
> **No changes on CI** But it works locally. Perhaps problem is related to CI launcher - CI has no active GUI because it runs as service.

Hi @alalek , 

Thanks for your feedback. I guess CI enviroment cannot fully support MSMF. BTW, you're referenceing to an error (i.e. MFVideoFormat_RGB32) which seems to be irrelevant to my recent HEVC/H264 FourCC code change. 

I only saw 2 failed tests from here (https://pullrequest.opencv.org/buildbot/builders/precommit_custom_windows/builds/100080/steps/test_videoio/logs/tests%20summary),

Should I remove HEVC/H265 from the "videoio_synthetic" unit test, or any suggestion to move forward?

Thank you,
Lilo

File: modules/videoio/test/test_video_io.cpp [hunk 2]
```diff
@@ -388,6 +388,7 @@ static Ext_Fourcc_PSNR synthetic_params[] = {
     {"wmv", "WMV3", 30.f, CAP_MSMF},
     {"wmv", "WVC1", 30.f, CAP_MSMF},
     {"mov", "H264", 30.f, CAP_MSMF},
+    {"mov", "HEVC", 30.f, CAP_MSMF},
```
Comment (alalek) [inline modules/videoio/test/test_video_io.cpp:391]: Please comment these lines:

```
    // {"mov", "HEVC", 30.f, CAP_MSMF},  // excluded due to CI issue: https://github.com/opencv/opencv/pull/23172
```

**Review by alalek — APPROVED** [review]
Well done! Thank you for the contribution :+1:

**Comment by lilohuang** [comment]
> Well done! Thank you for the contribution 👍

Thank you :)

**Comment by TrueWodzu** [comment]
@lilohuang I am want to tray that (and report back), Is it enough to build latest OpenCV with WITH_FFMPEG and OPENCV_ENABLE_NONFREE or do I have to download custom ffmpeg libraries and somehow point to them during CMake configuration?

**Comment by lilohuang** [comment]
> @lilohuang I want to tray that (and report back), Is it enough to build latest OpenCV with WITH_FFMPEG and OPENCV_ENABLE_NONFREE or do I have to download custom ffmpeg libraries and somehow point to them during CMake configuration?

This PR is *nothing* to do with FFMPEG, it's all about MSMF (Microsoft Media Foundation). You need to enable WITH_MSMF and compile OpenCV with Windows 10 SDK to take the benefit of [Microsoft Media Foundation H265 encoder ](https://learn.microsoft.com/en-us/windows/win32/medfound/h-265---hevc-video-encoder). Good luck!

**Pull request merged**